File: arrayADT.py
#LN04배열ADT_Python


#함수
def insert(array, data):
    global cur_pos
    if not array:
        for i in range(0,len(data),1):
            if data[i]=='+':
                array.append(data[i+1])
        cur_pos=len(array)-1
    else:
        for i in range(0,len(data),1):
            if data[i]=='+':
                array[cur_pos+1:cur_pos+1]=data[i+1]
                cur_pos+=1
        
        cur_pos=array.index(data[i])
        

def traverse_front(array):
    global cur_pos
    cur_pos=0
    

def traverse_rear(array):
    global cur_pos
    cur_pos=len(array)-1


def prev(array):
    global cur_pos
    cur_pos-=1


def next(array):
    global cur_pos
    cur_pos+=1


def delete(array):
    global cur_pos
    array.remove(array[cur_pos])
    if len(array)-1 < cur_pos:
        cur_pos-=1
    

def get_data(array):
    global cur_pos
    print('Return : %c\n'%array[cur_pos],end='')


def replace(array, new_data):
    global cur_pos
    for i in range(0,len(new_data),1):
            if new_data[i]=='=':
                array[cur_pos]=(new_data[i+1])
        
def empty(array):
    array.clear()
    print('empty array\n',end='')
    

def move(array, new_position):
    global cur_pos
    if new_position<cur_pos:
        array.insert(new_position,array[cur_pos])
        array.pop(cur_pos+1)
        cur_pos=new_position
    elif new_position>cur_pos:
        array.insert(new_position+1,array[cur_pos])
        array.pop(cur_pos)
        cur_pos=new_position
    else:
        print('현재 위치입니다.')
        

def move_front(array):
    global cur_pos
    array.insert(0, array[cur_pos])
    array.pop(cur_pos+1)
    cur_pos=0
        

def move_end(array):
    global cur_pos
    array.insert(len(array), array[cur_pos])
    array.pop(cur_pos)
    cur_pos=len(array)-1
        

def move_prev(array):
    global cur_pos
    array.insert(cur_pos-1, array[cur_pos])
    array.pop(cur_pos+1)
    cur_pos-=1
        


def move_next(array):
    global cur_pos
    array.insert(cur_pos+2, array[cur_pos])
    array.pop(cur_pos)
    cur_pos+=1
        
def find(array,f_data):
    print('검색하신 문자의 인덱스는 %d입니다.'%array.index(f_data))


def sortup(array):
    array.sort()
    

def sortlow(array):
    array.sort(reverse=True)

    
def help():
    print('='*14+'H E L P'+'='*14)
    print('{:22}{}'.format('INSERT',': +(data)'))
    print('{:22}{}'.format('DELETE',': -'))
    print('{:22}{}'.format('DELETE LAST',': > -'))
    print('{:22}{}'.format('TRAVERSE_FRONT',': <'))
    print('{:22}{}'.format('TRAVERSE_REAR',': >'))
    print('{:22}{}'.format('GET_DATA',': @'))
    print('{:22}{}'.format('REPLACE',': =(data)'))
    print('{:22}{}'.format('EMPTY',': E'))
    print('{:22}{}'.format('MOVE',': M(position)'))
    print('{:22}{}'.format('MOVE FRONT',': Mf'))
    print('{:22}{}'.format('MOVE END',': Mn'))
    print('{:22}{}'.format('MOVE PREV',': MP'))
    print('{:22}{}'.format('MOVE NEXT',': MN'))
    print('{:22}{}'.format('PRINT',': L'))
    print('{:22}{}'.format('PREV',': P'))
    print('{:22}{}'.format('NEXT',': N'))
    print('{:22}{}'.format('FIND',': F(data)'))
    print('{:22}{}'.format('SORTUP',': SU'))
    print('{:22}{}'.format('SORTLOW',': SL'))
    print('{:22}{}'.format('HELP',': H'))
    print('{:22}{}'.format('QUIT',': Q'))
    print()
    print('{:22}'.format('**Spacing between commands'))
    print('{:22}'.format('**Only one character for one data'))
    print('='*35)
    

#변수
my_array=[]
global cur_pos

help()

while True:

    
    
    key = list(input('>> '))

    if 'E' in key:
        empty(my_array)


    elif 'H' in key:
        print()
        help()

    elif 'F' in key:
        while 'F' in key:
            key.remove('F')
            data=key[0]
            find(my_array, data)

    elif 'Q' in key:
        print('프로그램을 종료합니다.')
        break

    else:
        
        if '+' in key:
            data=key
            insert(my_array,data)

        elif '@' in key:
            get_data(my_array)

        

        elif '-' in key:
            if '>' in key and '-' in key:
                key.remove('>')
                key.remove('-')
                traverse_rear(my_array)
                delete(my_array)
                cur_pos=0
                for i in range(0,len(key),1):
                    if key[i]=='-':
                        delete(my_array)  
            else:
                delete(my_array)

            
        elif '<' in key:
            if '<' in key and 'N' in key:
                traverse_front(my_array)
                for i in range(0,len(key),1):
                    if key[i]=='N':
                        next(my_array)
            elif '<' in key and 'M' in key:
                while '<' in key:
                    key.remove('<')
                while 'M' in key:
                    key.remove('M')
                while ' ' in key:
                    key.remove(' ')
                position=int(key[0])
                traverse_front(my_array) 
                move(my_array, position)
                
            else:
                traverse_front(my_array)    

            
        elif '>' in key:
            if '>' in key and '-' in key:
                key.remove('>')
                key.remove('-')
                traverse_rear(my_array)
                delete(my_array)
                cur_pos=0
                for i in range(0,len(key),1):
                    if key[i]=='-':
                        delete(my_array)
            elif '>' in key and 'P' in key:
                traverse_rear(my_array)
                for i in range(0,len(key),1):
                    if key[i]=='P':
                        prev(my_array)    
            else:  
                traverse_rear(my_array)


        elif '-' in key:
            delete(my_array)

        

        elif '=' in key:
            data=key
            replace(my_array, data)
            
            
        elif 'M' in key:
            if 'M' in key and 'f' in key:
                move_front(my_array)
            elif 'M' in key and 'n' in key:
                move_end(my_array)
            elif 'M' in key and 'P' in key:
                move_prev(my_array)
            elif 'M' in key and 'N' in key:
                move_next(my_array)
            else:
                while 'M' in key:
                    key.remove('M')
                position=int(key[0])
                move(my_array, position)

        elif 'P' in key:
            prev(my_array)
        
        elif 'N' in key:
            next(my_array)


        elif 'S' in key:
            if 'S' in key and 'U' in key:
                sortup(my_array)
            elif 'S'in key and 'L' in key:
                sortlow(my_array)

        elif 'L' in key:
            my_array[cur_pos]='('+my_array[cur_pos]+')'
            for i in range(0,len(my_array)):
                print("%2s"%my_array[i], end=' ')
            my_array[cur_pos]=my_array[cur_pos].replace('(','')
            my_array[cur_pos]=my_array[cur_pos].replace(')','')
            print()
            continue

        my_array[cur_pos]='('+my_array[cur_pos]+')'
        for i in range(0,len(my_array)):
            print("%2s"%my_array[i], end=' ')
        my_array[cur_pos]=my_array[cur_pos].replace('(','')
        my_array[cur_pos]=my_array[cur_pos].replace(')','')
        print()

    
  

File: arrayADT/arrayADT/arrayADT.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#define NUM 30
#pragma warning(disable : 4996)


//함수
void traverse_front(char* arr);
void traverse_rear(char* arr);
void prev(char* arr);
void next(char* arr);
void delete(char* arr);
void get_data(char* arr);
void replace(char* arr);
void empty(char* arr);
void move(char* arr, int position);
void move_front(char* arr);
void move_end(char* arr);
void move_prev(char* arr);
void move_next(char* arr);
void find(char* arr, int position);
void odd(char* arr);
void even(char* arr);
void help();



//변수
int cur_pos = -1; //현재 위치
int position;
char new_data;
char arr[NUM]; //정보가 저장된 배열
char user[NUM]; //사용자로부터 입력받은 배열


//main 함수
int main(void) {
	char arr[NUM] = { '\0' };
	char user[NUM] = { '\0' };
	int j = 0;

	help();

	while (1) {
		printf("\n>> ");
		gets(user);



		if (strstr(user, "E") != NULL) {
			empty(arr);
			continue;
		}

		else if (strstr(user, "F") != NULL) {
			for (int i = 0; i < NUM; i++) {
				if (user[i] == 'F') {
					user[i] = user[i + 1];
					user[i + 1] = '\0';
				}
			}
			char c = user[0];
			int n = 0;
			n = c - '0';
			find(arr, n);
			continue;
		}

		else if (strstr(user, "O") != NULL) {
			odd(arr);
			continue;
		}

		else if (strstr(user, "V") != NULL) {
			even(arr);
			continue;
		}

		else if (strstr(user, "H") != NULL) {
			printf("\n");
			help();
			continue;
		}

		else if (strstr(user, "Q") != NULL) {
			printf("프로그램을 종료합니다.\n");
			break;
		}

		else {


			if (strstr(user, "+") != NULL) {
				if (arr[0] == NULL) {
					for (int i = 0; i < NUM; i++) {
						if (user[i] == '+') {
							arr[cur_pos + 1] = user[i + 1];
							cur_pos++;
						}
					}
				}
				else {
					for (int i = 0; i < NUM; i++) {
						if (user[i] == '+') {
							arr[cur_pos + 1] = user[i + 1];
							cur_pos++;
						}
						arr[cur_pos + 1] = arr[cur_pos];
					}
					arr[cur_pos + 1] = '\0';
					cur_pos -= 1;
				}
			}


			else if (strstr(user, "@") != NULL) {
				get_data(arr);
			}

			else if ((strstr(user, ">") != NULL) && (strstr(user, "-") != NULL)) {
				traverse_rear(arr);
				delete(arr);
				cur_pos = 0;
				for (int i = 3; i < NUM; i++) {
					if (user[i] == '-') {
						delete(arr);
					}
				}
			}

			else if (strstr(user, "-") != NULL) {
				delete(arr);
			}

			else if ((strstr(user, "<") != NULL) && (strstr(user, "N") != NULL)) {
				traverse_front(arr);
				for (int i = 0; i < NUM; i++) {
					if (user[i] == 'N') {
						next(arr);
						i++;
					}
				}
			}

			else if ((strstr(user, "<") != NULL) && (strstr(user, "M") != NULL)) {
				traverse_front(arr);
				for (int i = 0; i < NUM; i++) {
					if (user[i] == 'M') {
						user[i] = user[i + 1];
						user[i + 1] = '\0';
					}
				}
				char c = user[2];
				int n = 0;
				n = c - '0';
				move(arr, n);
			}

			else if (strstr(user, "<") != NULL) {
				traverse_front(arr);
			}

			else if ((strstr(user, ">") != NULL) && (strstr(user, "P") != NULL)) {
				traverse_rear(arr);
				for (int i = 0; i < NUM; i++) {
					if (user[i] == 'P') {
						prev(arr);
						i++;
					}
				}
			}

			else if (strstr(user, ">") != NULL) {
				traverse_rear(arr);
			}


			else if (strstr(user, "=") != NULL) {// user 문자열에 =이 포함되었는지 확인
				for (int i = 0; i < NUM; i++) {
					if (user[i] == '=') {
						arr[cur_pos] = user[i + 1];
					}
				}
			}

			else if (strstr(user, "M") != NULL) {// user 문자열에 =이 포함되었는지 확인

				if (strstr(user, "f") != NULL) {
					move_front(arr);
				}
				else if (strstr(user, "n") != NULL) {
					move_end(arr);
				}
				else if (strstr(user, "P") != NULL) {
					move_prev(arr);
				}
				else if (strstr(user, "N") != NULL) {
					move_next(arr);
				}

				else {
					for (int i = 0; i < NUM; i++) {
						if (user[i] == 'M') {
							user[i] = user[i + 1];
							user[i + 1] = '\0';
						}
					}
					char c = user[0];
					int n = 0;
					n = c - '0';
					move(arr, n);
				}

			}

			else if (strstr(user, "P") != NULL) {
				prev(arr);
			}

			else if (strstr(user, "N") != NULL) {
				next(arr);
			}

			else if (strstr(user, "L") != NULL) {
				for (int i = 0; i < strlen(arr); i++) {
					if (i == cur_pos) {
						printf("(%c) ", arr[i]);
					}
					else {
						printf("%c ", arr[i]);
					}
				}
				continue;
			}

		}

		for (int i = 0; i < strlen(arr); i++) {
			if (i == cur_pos) {
				printf("(%c) ", arr[i]);
			}
			else {
				printf("%c ", arr[i]);
			}
		}
	}

	return 0;

}


void traverse_front(char* arr) {
	cur_pos = 0;
}


void traverse_rear(char* arr) {
	cur_pos = strlen(arr) - 1;
}


void prev(char* arr) {
	cur_pos -= 1;
}


void next(char* arr) {
	cur_pos += 1;
}


void delete(char* arr) {
	for (int i = cur_pos; i < strlen(arr); i++) {
		arr[i] = arr[i + 1];
	}
	if (strlen(arr) - 1 < cur_pos)
		cur_pos -= 1;
}


void get_data(char* arr) {
	printf("Return : %c\n", arr[cur_pos]);
}


void replace(char* arr) {
	char user[NUM] = { '\0' };
	for (int i = 0; i < NUM; i++) {
		if (user[i] == '=') {
			arr[cur_pos] = user[i + 1];
		}
	}
}


void empty(char* arr) {
	for (int i = 0; i < strlen(arr); i++)
		arr[i] = '\0';
	printf("empty array");
	cur_pos = -1;
}


void move(char* arr, int position) {
	if (position < cur_pos) {
		char a;
		a = arr[cur_pos];
		for (int i = cur_pos; i > position; i--) {
			arr[i] = arr[i - 1];
		}
		arr[position] = a;
		cur_pos = position;
	}
	else if (position > cur_pos) {
		char a;
		a = arr[cur_pos];
		for (int i = cur_pos; i < position + 1; i++) {
			arr[i] = arr[i + 1];
		}
		arr[position] = a;
		cur_pos = position;
	}
	else
		printf("현재 위치입니다.");
}


void move_front(char* arr) {
	char a;
	a = arr[cur_pos];
	for (int i = cur_pos; i > 0; i--) {
		arr[i] = arr[i - 1];
	}
	arr[0] = a;
	cur_pos = 0;
}


void move_end(char* arr) {
	char a;
	a = arr[cur_pos];
	for (int i = cur_pos; i < strlen(arr) + 1; i++) {
		arr[i] = arr[i + 1];
	}
	arr[strlen(arr)] = a;
	cur_pos = strlen(arr) - 1;
}


void move_prev(char* arr) {
	char a;
	a = arr[cur_pos];
	arr[cur_pos] = arr[cur_pos - 1];
	arr[cur_pos - 1] = a;
	cur_pos -= 1;
}


void move_next(char* arr) {
	char a;
	a = arr[cur_pos];
	arr[cur_pos] = arr[cur_pos + 1];
	arr[cur_pos + 1] = a;
	cur_pos += 1;
}

void find(char* arr, int position) {
	printf("검색한 인덱스의 값: %c ", arr[position]);

}


void odd(char* arr) {
	printf("홀수 인덱스: ");
	for (int i = 0; i < strlen(arr); i++) {
		if (i % 2 == 1) {
			printf("%c ", arr[i]);
		}
	}
}

void even(char* arr) {
	printf("짝수 인덱스(+0): ");
	for (int i = 0; i < strlen(arr); i++) {
		if ((i + 2) % 2 == 0) {
			printf("%c ", arr[i]);
		}
	}
}


void help() {
	for (int i = 0; i < 15; i++) {
		printf("-");
	}
	printf("H E L P");
	for (int i = 0; i < 15; i++) {
		printf("-");
	}
	printf("\n%-22s %s\n", "INSERT", ": +(data)");
	printf("%-22s %s\n", "DELETE", ": -");
	printf("%-22s %s\n", "DELETE LAST", ": > -");
	printf("%-22s %s\n", "TRAVERSE_FRONT", ": <");
	printf("%-22s %s\n", "TRAVERSE_REAR", ": >");
	printf("%-22s %s\n", "GET_DATA", ": @");
	printf("%-22s %s\n", "REPLACE", ": =(data)");
	printf("%-22s %s\n", "EMPTY", ": E");
	printf("%-22s %s\n", "MOVE", ": M(position)");
	printf("%-22s %s\n", "MOVE FRONT", ": Mf");
	printf("%-22s %s\n", "MOVE END", ": Mn");
	printf("%-22s %s\n", "MOVE PREV", ": MP");
	printf("%-22s %s\n", "MOVE NEXT", ": MN");
	printf("%-22s %s\n", "PRINT", ": L");
	printf("%-22s %s\n", "PREV", ": P");
	printf("%-22s %s\n", "NEXT", ": N");
	printf("%-22s %s\n", "FIND", ": F(position)");
	printf("%-22s %s\n", "ODD", ": O");
	printf("%-22s %s\n", "EVEN", ": V");
	printf("%-22s %s\n", "HELP", ": H");
	printf("%-22s %s\n", "QUIT", ": Q");
	printf("\n");
	printf("%-22s\n", "**Spacing between commands");
	printf("%-22s\n", "**Only one character for one data");
	for (int i = 0; i < 35; i++) {
		printf("-");
	}


}